### PR TITLE
Fix font feature tag validation

### LINKED ISSUE #13517: [Windows] When adding font_features to settings, zed crashes until settings removed
### Check for existing issues

- [X] Completed

## Description
Adding font features (`ui_font_features` | `buffer_font_features`) causes a crash.

tested properties:
- `italic` and `underlined` crashes
- example on the [zed docs](https://zed.dev/docs/configuring-zed#buffer-font-features) does not

## Reproduce
- open settings
- add e.g. `"buffer_font_features": {"italic":true}`
- zed crashes

## Error message when using "cargo run" with RUST_BACKTRACE
[error_msg.txt](https://github.com/user-attachments/files/15974543/error_msg.txt)

### Environment

OS: Windows 11
Processor: Intel(R) Core(TM) i7-14700HX   2.10 GHz
RAM: 32GB

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>expand for logs...</summary>
//When starting zed after crash
2024-06-25T18:15:46+02:00 [INFO] ========== starting zed ==========
2024-06-25T18:15:46+02:00 [INFO] Use Segoe UI as UI font.
2024-06-25T18:15:46+02:00 [INFO] Using direct write text system.
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Using git binary path: None
2024-06-25T18:15:46+02:00 [ERROR] crates\zed\src/main.rs:362: SHELL environment variable is not assigned so we can't source login environment variables

Caused by:
    environment variable not found
2024-06-25T18:15:46+02:00 [INFO] extensions updated. loading 6, reloading 0, unloading 0
2024-06-25T18:15:46+02:00 [ERROR] crates\zed\src\zed.rs:716: Expected array of length 2
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] set status on client 0: Authenticating
2024-06-25T18:15:46+02:00 [INFO] set status on client 177145: Connecting
2024-06-25T18:15:46+02:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-06-25T18:15:46+02:00 [INFO] Enabling Vulkan Portability
2024-06-25T18:15:46+02:00 [INFO] Enabling color space support
2024-06-25T18:15:46+02:00 [INFO] Adapter "NVIDIA GeForce RTX 4060 Laptop GPU"
2024-06-25T18:15:46+02:00 [INFO] Ray tracing is supported
2024-06-25T18:15:46+02:00 [WARN] Requested size 1x1 is outside of surface capabilities
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-06-25T18:15:46+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-06-25T18:15:46+02:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_impl")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_transformed")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("blend_color")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_impl")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_device_position_transformed")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("to_tile_position")
2024-06-25T18:15:46+02:00 [INFO] Skip function Some("blend_color")
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:46+02:00 [ERROR] crates\fs\src/fs.rs:537: Error { kind: Generic("Input watch path is neither a file nor a directory."), paths: [] }
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Opening main db
2024-06-25T18:15:46+02:00 [INFO] Using Windows API for pseudoconsole
2024-06-25T18:15:47+02:00 [INFO] Using surface present mode MAILBOX
2024-06-25T18:15:47+02:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
</details>

### PR DESCRIPTION
**Pull request by JunkuiZhang**
The previous implementation that I implemented had two issues:
1. It did not throw an error when the user input some invalid values such as "panic".
2. The feature tag for OpenType fonts should be a combination of letters and digits. We only checked if the input was an ASCII character, which could lead to undefined behavior.

Closes #13517 

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/text_system/font_features.rs [hunk 1]
```diff
@@ -142,3 +142,19 @@ impl schemars::JsonSchema for FontFeatures {
...
 }
+
+fn is_valid_feature_tag(tag: &str) -> bool {
+    let raw_bytes = tag.bytes();
+    if raw_bytes.len() != 4 {
+        return false;
+    }
+    for ch in raw_bytes.into_iter() {
+        match ch {
+            48..=57 => continue,  // 0 ... 9
```
Comment (maxdeviant) [inline crates/gpui/src/text_system/font_features.rs:153]: Seems like we could operate on `char`s instead to make these match arms a bit more self-explanatory?

```suggestion
            '0'..='9' => continue,  // 0 ... 9
```

Comment (JunkuiZhang) [reply crates/gpui/src/text_system/font_features.rs]: Thanks! Changes updated.

File: crates/gpui/src/text_system/font_features.rs [hunk 2]
```diff
@@ -142,3 +142,18 @@ impl schemars::JsonSchema for FontFeatures {
...
     }
 }
+
+fn is_valid_feature_tag(tag: &str) -> bool {
+    if tag.len() != 4 {
+        return false;
+    }
+    for ch in tag.chars() {
+        match ch {
+            '0'..='9' => continue,
```
Comment (maxdeviant) [inline crates/gpui/src/text_system/font_features.rs:152]: I think we can replace this `match` with `ch.is_ascii_alphanumeric`: https://doc.rust-lang.org/std/primitive.char.html#method.is_ascii_alphanumeric

Comment (JunkuiZhang) [reply crates/gpui/src/text_system/font_features.rs]: It's surprising that Rust already has this function.

**Review by maxdeviant — APPROVED** [review]
Thanks!

**Pull request merged**